[clang] Implement CTAD for type alias template.

hokein opened the pull request:
Fixes #54051

This patch implements the C++20 feature -- CTAD for alias templates (P1814R0, specified in https://eel.is/c++draft/over.match.class.deduct#3). It is an initial patch:
- it cover major pieces, thus it works for most cases;
- the big missing piece is to implement the associated constraints (over.match.class.deduct#3.3) for the synthesized deduction guides, see the FIXME in code and tests;
- Some enhancements on the TreeTransform&TemplateInstantiator to allow performing instantiation on `BuildingDeductionGuides` mode;

cor3ntin commented:
I'm thrilled to see work in this area. Thanks!

Note that Clang 18 is going to branch soon, so this PR is going to target Clang 19.
However, because it took a while to stabilize other CTAD related works, i think it might be worth it to try to complete that work early in the Clang 19 cycle, so that we have a few months to weed out the bugs.

I do believe this is the last unimplemented features for our C++20 support (outside of bugs and modules)

hokein commented:
> Note that Clang 18 is going to branch soon, so this PR is going to target Clang 19.

Yes, this patch will not be ready before the 18 release cut, so targeting clang 19 sounds like a good plan.

As you may notice, the implementation is incomplete, and is likely far from perfect, I can see two alternatives to proceed:
1) we keep improving it here, and landing it when it is in good/perfect shape;
2) find a way to land a relatively-good initial version, and continue to develop it incrementally;

Re 1), it is hard to predict the timeline, I'd like to finish it soon, but I also have other tasks with higher priority.
2) seems like a good option, we can hide this experimental feature with a new `-cc1` flag, when this feature is ready, we enable it by default. It is probably not a common practice for a narrow feature like this. 

Any thoughts?  @cor3ntin, @shafik, @AaronBallman

hokein committed:
[clang] Implement Class Template Argument Deduction (CTAD) for type alias
templates P1814R0.

This patch implements the C++20 feature -- CTAD for alias templates.
This is an initial patch, which covers most of pieces, the major missing
piece is to implement the associated constraints (over.match.class.deduct#3.3)
for the synthesized deduction guides (we can address in a followup).

This patch also refactors the existing `ConvertConstructorToDeductionGuideTransform`
to allow code reuse.

cor3ntin commented:
@hokein @sam-mccall to repeat what was said in the language group last week.

We do not think a feature flag is a good fit. 

 - Until clang 19 ships (in 6 months) we do not need to stabilize the feature (but ofc we should avoid regressions, which we would not find if hidden behind a flag), and the flag does not guarantee existing features are not impacted.
 
And if we break trunk, we revert, you get extra test cases... it's perfectly fine

 - The feature should be behind a -std=C++20 flag.
 - We should not update __cpp_deduction_guides until the feature is battle tested

We do that for about every feature that is not concept/module scale (and then again, these things were tses), and this isn't big enough to warrant a novel approach.
  
I still think we should land this soon and we can start a review whenever you are ready.
If you want to land it in an incomplete state, we just need to make sure we have a good collective understanding of the remaining work that would need to be done to bring the feature to completion (tests with fix me, issues, etc)

Does that make sense to you?

hokein commented:
Thanks for the summary.

> We do not think a feature flag is a good fit.
> 
> * Until clang 19 ships (in 6 months) we do not need to stabilize the feature (but ofc we should avoid regressions, which we would not find if hidden behind a flag), and the flag does not guarantee existing features are not impacted.
> 

I think this works fine with clang open-source release cycles (18 release was just cut, we have 6 months from now to polish this feature). However, this is not applicable for our internal case. At Google, we have used C++20 in production, and we have our own toolchain release cycles (which stay close to upstream main). Checking in an in-development C++20 feature without a proper safeguard poses risks to users (clang crashes, incorrect compilations, and other issues).

So having a temporary flag is useful, it enables us to incrementally test and experiment it without compromising the experience for most users. 

> And if we break trunk, we revert, you get extra test cases... it's perfectly fine

We could revert the whole patch when things are broken (and it is what we usually do). But I'm not sure revert/reland a big patch back and forth is a good idea.

This also brings a general question: what level of quality should we require from experimental features in HEAD?

> We do that for about every feature that is not concept/module scale (and then again, these things were tses), and this isn't big enough to warrant a novel approach.

Yeah, I agree that this is a narrow feature. Given that this feature is very late to the party, and C++20 has been used in some productions, I think it is probably reasonable to have a flag.

> I still think we should land this soon and we can start a review whenever you are ready.
If you want to land it in an incomplete state, we just need to make sure we have a good collective understanding of the remaining work that would need to be done to bring the feature to completion (tests with fix me, issues, etc)

+1, I think the current state should be good enough and ready for review. The remaining big piece (see the FIXME) is to implement the associated constraints (over.match.class.deduct#3.3).

hokein commented:
I started a discussion on discourse: https://discourse.llvm.org/t/a-temporary-flag-for-guarding-access-to-an-in-development-c-20-feature-in-clang/76597.

shafik reviewed (commented):
In order to aid the discussion on the temporary flag, I think it would be helpful to have a full set of tests that show what elements won't work right with the current set of changes.

I think this will help us gauge how impactful partial support would be and maybe help guide the discussion on alternatives.

erichkeane commented:
>We could revert the whole patch when things are broken (and it is what we usually do). But I'm not sure revert/reland a big patch back and forth is a good idea.

This is effectively what we did repeatedly with the deferred concepts patch and IMO, was a really nice healthy way of doing it.  If you organize your commits (much easier on github than it was on Phab), you can present JUST the changes between revisions/fixes and get a really nice revert/fix/release cycle going.

I'd encourage just enabling by default (to avoid losing the test breadth that comes along with it), and reverting if there are sufficiently 'bad' breakages.  

Note that you'd not have to revert for all 'bugs', just those that break existing code.  We can hold off on setting the Feature Test Macro until we're comfortable with it though.

yronglin commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10636,6 +10889,113 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
   Guides.suppressDiagnostics();
 
+  SmallVector<DeclAccessPair> GuidesCandidates;
+  if (AliasTemplate) {
```
I was wondering if  we need to separate the code that tweaks the candidate set into another function.

hokein replied:
Yes, and this is not the proper place to construct deduction guides for alias templates (they should live in `DeclareImplicitDeductionGuides`).

hokein committed:
Rewrite the implementations:

- Move the implementation to SemaTemplate.cpp where the typical CTAD
  implementation lives, thus no sepatate CTAD.h/.cpp;
- Rewrite the implementation, we leverage more on clang's template
  instantatiation mechanism to build the deduction guide, it
  implementation is simplier;
- Some enhancements on TreeTransform and TemplateInstantiator to allow
  running on BuildingDeductionGuides mode;
- Added more tests;

hokein commented:
This patch is ready for review now.

cor3ntin reviewed (commented):
Looks great generally.

I did leave a few preliminary comments.
The alias ctad mechanism is not easy to get my head around so this is going to take a few rounds

Thanks!

hokein committed:
Address review comments, and add more tests.

hokein commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1446,6 +1447,58 @@ namespace {
+          TemplateArgumentLoc Input = SemaRef.getTrivialTemplateArgumentLoc(
+              pack, QualType(), SourceLocation{});
+          TemplateArgumentLoc Output;
+          if (!SemaRef.SubstTemplateArgument(Input, TemplateArgs, Output))
+            TArgs.push_back(Output.getArgument());
```
Good catch. Added the bailout.

hokein reviewed (commented):
Thanks for the initial review.

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2612,44 +2671,309 @@ struct ConvertConstructorToDeductionGuideTransform {
+            RT->getAsCXXRecordDecl())) {
+      Template = CTSD->getSpecializedTemplate();
+      AliasRhsTemplateArgs = CTSD->getTemplateArgs().asArray();
+    }
```
An `assert(false)` may make our lives easier in the future,

hokein replied:
Done.

hokein committed:
Bailout when SubstTemplateArgument fails.

hokein committed:
Add assertion on unhandled RHS type of the type alias.

erichkeane commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10598,10 +10598,36 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
+      LookupTemplateDecl = AliasTemplate;
+      auto UnderlyingType = AliasTemplate->getTemplatedDecl()
+                                ->getUnderlyingType()
+                                .getDesugaredType(Context);
```
Do we mean Desugar here, or Canonical?

hokein replied:
Here we'd like to get the underlying type with all alias stripped. I think both `Desugar` and `Canonical` can achieve the goal here, but `Desugar` seems more explicit.

erichkeane replied:
`Desugar` doesn't remove all of the typedefs/etc, and leaves sugar in pointees/references.  So this is very likely to end up being an underlying type of, for example, `SomeTypedef*`.  Canonical canonicalizes EVERYTHING.

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2258,6 +2258,94 @@ class ExtractTypeForDeductionGuide
+          : std::nullopt);
+  if (const auto *TC = TTP->getTypeConstraint())
+    SemaRef.SubstTypeConstraint(NewTTP, TC, Args,
+                                /*EvaluateConstraint*/ true);
```
```suggestion
                                /*EvaluateConstraint=*/ true);
```
Do we mean to evaluate the constraint?  Typically we don't want to evaluate the constraint unless we are checking the decl for a call, or for constraint comparisons.  So we really don't ever substitute these ever (and the substituted versions don't really end up in the AST).

hokein replied:
Yes, I think so, though the function parameter name is a bit confusing.

Here, we intend to *rewrite* the `TTP->getTypeConstraint()` (update the old template occurrences with the new template parameters). Looking at the `SubstTypeConstraint` implementation, if we set it to false, it will [attach](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaTemplateInstantiate.cpp#L2886-L2890) the old constraint to `NewTTP`.

(This is a refactoring change which lifts the existing function from `ConvertConstructorToDeductionGuideTransform` for code sharing purpose).

hokein replied:
> For the most part, we WANT the 'old' constraint to attach. All the work to instantiate the constraints end up going against the original template.

I think this is not we want for this case, we're in the process of building a new `FunctionTemplate` from the original template here, specifically building the `TemplateTypeParmDecl` for the new template. Attaching the `old` constraint to it is not correct because this will make the instantiation of the new template go against the template parameter in original template, and will lead to crashes in the normal CTAD cases, e.g. 

```
template <typename A, typename T>
concept True = true;
struct concepts {
  template <class Y>
  struct B {
    template <True<Y> Z> B(Y, Z);
  };
};

concepts::B b(0, 1);
```

erichkeane replied:
Hmm... but as-written it seems like it would cause problems where the original function-template isn't instantiate-able.  It seems to me that instead of substituting it, perhaps we should be 'converting' it in some way to update the references?  I'll have to think on it, though I'm @zygoloid  can come around and answer for us.

hokein replied:
This case is special, we're performing a template substitution with dependent template args. I think it is fine, the constraint satisfaction check is [not performed](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaTemplate.cpp#L5029) for dependent args.

erichkeane replied:
Hmm... if it is always dependent template args we're putting in that part should be fine.  At that point it becomes just a 'replace args' instead of a substitution, which should be ok.

However, are we making sure we're replacing them 1:1 (including parent templates)?  Else the depth of some of them is going to 'change' and later substitution is going to fail.

hokein replied:
Yeah, we're replacing them 1:1.

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2612,44 +2669,313 @@ struct ConvertConstructorToDeductionGuideTransform {
+
+bool hasDeclaredDeductionGuides(DeclarationName Name, DeclContext *DC) {
+  // Check whether we've already declared deduction guides for this template.
+  // FIXME: Consider storing a flag on the template to indicate this.
```
The cost of a lookup can be pretty bad, I think the flag on the template is worth doing ASAP.

hokein replied:
I think the flag change is orthogonal to the current patch, and we already do this lookup for non-type-alias CTAD implementation. 

We can address it in a followup.

erichkeane reviewed (commented):
We probably need an entry in ReleaseNotes.rst here as well.

hokein commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10598,10 +10598,36 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
   auto *Template =
       dyn_cast_or_null<ClassTemplateDecl>(TemplateName.getAsTemplateDecl());
+  TemplateDecl* LookupTemplateDecl = Template;
+  if (!Template && getLangOpts().CPlusPlus20) { // type alias template
```
We could add one for that, but I'm not sure, it seems that in general clang doesn't provide diagnostics for features that aren't available in the current `-std=` version (at least I didn't such a diagnostic).

For example, when we use C++ concepts without `-std=c++20`, clang rejects it with a diagnostic `unknown type name 'concept'`. gcc have better diagnostics (`note: 'concept' only available with '-std=c++20' or '-fconcepts'`).

I have a somewhat-relevant FIXME below, it is about to adjust `err_deduced_non_class_template_specialization_type` message to include the type-alias templates.

erichkeane replied:
We actually have a TON of those sorts of things: See the warning groups named like `CXXPre20Compat`.

hokein commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10598,10 +10598,36 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
       dyn_cast_or_null<ClassTemplateDecl>(TemplateName.getAsTemplateDecl());
+  TemplateDecl* LookupTemplateDecl = Template;
+  if (!Template && getLangOpts().CPlusPlus20) { // type alias template
+    if (auto *AliasTemplate = dyn_cast_or_null<TypeAliasTemplateDecl>(
```
oh, good catch (this part of code has been moved and rewritten several times, I probably missed to cleanup it). Done.

hokein committed:
Address review comments:

- simplify the if
- add a release note

hokein committed:
- add a C++20-compat diagnostic warning
- add more tests

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2612,44 +2671,309 @@ struct ConvertConstructorToDeductionGuideTransform {
+// Find all template parameters of the AliasTemplate that appear in the
+// given DeducedArgs.
+SmallVector<unsigned>
+FindAppearedTemplateParamsInAlias(ArrayRef<TemplateArgument> DeducedArgs,
```
@erichkeane

erichkeane replied:
I might suggest we instead genericize this to not be AliasTemplate specific, and have it take a list of NamedDecls and template args.  Then call it something like, `DeclsReferencedInTemplateArgumentList`.

I also probably wouldn't have this return a list of 'ints', and just return the pointers themselves, as this little trade back and forth is a little awkward.

hokein replied:
> I might suggest we instead genericize this to not be AliasTemplate specific, and have it take a list of NamedDecls and template args. Then call it something like, DeclsReferencedInTemplateArgumentList.

We could make this function more generic, but this function is only be used for alias-template CTAD implementation, and it is unlikely to be reused for other things, so the motivation to genericize it doesn't seem strong to me (happy to do this change if you feel strong about it).

> I also probably wouldn't have this return a list of 'ints', and just return the pointers themselves, as this little trade back and forth is a little awkward.

Yeah, I considered this in the initial version, but I found returning a list of Index can simplify the code. The caller of this function needs the `Index`, if we return a list of pointers, the caller needs to do some more work to figure out their corresponding Index.

erichkeane replied:
The generic suggestion is because it would, I believe, make this a much more readable function with very minor changes in the 'use'.  So I think there is value there.

I see what you mean about the index, I think it is alright to leave the index part.

hokein replied:
OK, that makes sense. I made it a generic function.

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2258,6 +2260,94 @@ class ExtractTypeForDeductionGuide
 
+// Build a deduction guide with the specified parameter types.
+FunctionTemplateDecl *
+buildDeductionGuide(Sema &SemaRef, TemplateDecl *OriginalTemplate,
+                    TemplateParameterList *TemplateParams,
+                    CXXConstructorDecl *Ctor, ExplicitSpecifier ES,
+                    TypeSourceInfo *TInfo, SourceLocation LocStart,
+                    SourceLocation Loc, SourceLocation LocEnd, bool IsImplicit,
+                    llvm::ArrayRef<TypedefNameDecl *> MaterializedTypedefs = {}) {
```
I missed that was pre existing, (but they are pretty bad names!)

hokein commented:
@cor3ntin, @erichkeane, thanks for all review comments. I'm not sure you have reviewed the whole patch. My feeling is that the core part (`DeclareImplicitDeductionGuidesForTypeAlias`) may not have received sufficient reviews. It would be great if you can take a closer look at this part.

erichkeane commented:
> @cor3ntin, @erichkeane, thanks for all review comments. I'm not sure you have reviewed the whole patch. My feeling is that the core part (`DeclareImplicitDeductionGuidesForTypeAlias`) may not have received sufficient reviews. It would be great if you can take a closer look at this part.

I've looked through it a few times(including again now) and didn't see anything that stuck out as wrong to me.  I don't believe I can review it any more thoroughly.

hokein committed:
Rename FindAppearedTemplateParamsInAlias => TemplateParamsReferencedInTemplateArgumentList

hokein commented:
> > @cor3ntin, @erichkeane, thanks for all review comments. I'm not sure you have reviewed the whole patch. My feeling is that the core part (`DeclareImplicitDeductionGuidesForTypeAlias`) may not have received sufficient reviews. It would be great if you can take a closer look at this part.
> 
> I've looked through it a few times(including again now) and didn't see anything that stuck out as wrong to me. I don't believe I can review it any more thoroughly.

Great, thanks. I think I have addressed all comments.

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -96,6 +96,10 @@ C++20 Feature Support
 
+- Initial support for class template argument deduciton (CTAD) for type alias
+  templates.
+  (`#54051 https://github.com/llvm/llvm-project/issues/54051`_).
```
```suggestion
  (#GH54051).
```

hokein replied:
This doesn't follow the pattern used in this file, we use `#<number> https://github.com/llvm/llvm-project/issues/<nummber>`.

cor3ntin replied:
That changes recently (either way, you were missing <>)

hokein replied:
I see, I were not aware of it. Switched to the `GH` form.

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2612,44 +2669,312 @@ struct ConvertConstructorToDeductionGuideTransform {
+          Context.getCanonicalTemplateArgument(
+              Context.getInjectedTemplateArg(NewParam));
+    }
+    // FIXME: support require clause.
```
So this is missing https://eel.is/c++draft/over.match.class.deduct#3.3
I think we need to create a Binary (and) op expression with the requires clause of F and, a literal bool expression set to whether the arguments are deducible (do we implement the deducible part?)

cor3ntin replied:
I think it's fine to leave that to a subsequent patch

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2612,44 +2669,312 @@ struct ConvertConstructorToDeductionGuideTransform {
+            F, TemplateArgListForBuildingFPrime, AliasTemplate->getLocation(),
+            Sema::CodeSynthesisContext::BuildingDeductionGuides)) {
+      auto *GG = dyn_cast<CXXDeductionGuideDecl>(FPrime);
+      // FIXME: implement the associated constraint per C++
+      // [over.match.class.deduct]p3.3:
+      //    The associated constraints ([temp.constr.decl]) are the
+      //    conjunction of the associated constraints of g and a
+      //    constraint that is satisfied if and only if the arguments
+      //    of A are deducible (see below) from the return type.
```
That FIXME should probably be merged with the one above

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1446,6 +1447,59 @@ namespace {
+      std::vector<TemplateArgument> TArgs;
+      switch (Arg.getKind()) {
+      case TemplateArgument::Pack:
+        // Iterially rewrite the template argument pack, instead of unpacking
```
Iteratively ?

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -6286,8 +6284,18 @@ NamedDecl *Sema::FindInstantiatedDecl(SourceLocation Loc, NamedDecl *D,
+          CXXRecordDecl *SubstRecord = T->getAsCXXRecordDecl();
+
+          if (!SubstRecord) {
+            // The T can be a dependent TemplateSpecializationType when
+            // performing a substitution for building a deduction guide,
```
```suggestion
            // T can be a dependent TemplateSpecializationType when
            // performing a substitution for building a deduction guide.
```

hokein committed:
Address review comments.

hokein commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10598,10 +10598,34 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
+        if (const auto *CTSD = llvm::dyn_cast<ClassTemplateSpecializationDecl>(
+                RT->getAsCXXRecordDecl()))
+          Template = CTSD->getSpecializedTemplate();
+      }
```
ah, sorry, I added it and then removed later. There are some other cases which are not supported, see the testcase `test17`.

hokein replied:
Yes, exactly.

hokein commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -2612,44 +2669,312 @@ struct ConvertConstructorToDeductionGuideTransform {
+SmallVector<unsigned> TemplateParamsReferencedInTemplateArgumentList(
+    ArrayRef<NamedDecl *> TemplateParams,
+    ArrayRef<TemplateArgument> DeducedArgs) {
+  struct FindAppearedTemplateParams
```
I missed this one when renaming the function. What do you think about the current one `ReferenceFinder`?

erichkeane replied:
> I missed this one when renaming the function. What do you think about the current one `ReferenceFinder`?

ReferenceFinder is a little overloaded of a name :)  Perhaps `TypeReferencedFinder`? Note the 'd' which I think prevents it from being mistaken with a "finds all things with an L or R value reference".

hokein replied:
OK, I'm not sure `Type` is the best fit, I'd stay close with template params.  Renamed it to `TemplateParamsReferencedFinder` (if we want to use a less-verbose name, I'm fine with `Finder` since this is a structure defined in a function body).

hokein commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -96,6 +96,10 @@ C++20 Feature Support
   (`#79240 <https://github.com/llvm/llvm-project/issues/79240>`_).
 
+- Initial support for class template argument deduciton (CTAD) for type alias
+  templates.
```
Done.

hokein commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10598,10 +10598,38 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
+    }
+  }
+  if (!Template) {
+    // FIXME: update the diagnostic message to include C++20 alias templates
```
Are you suggesting to address it now? My original plan is to address it in a followup, but it seems small enough to do it in this patch.

cor3ntin replied:
I think we might as well, yes (unless it's more complicated than i think it is)

hokein replied:
Sure, please take a look on the new change. I also rename the diagnostic to `err_deduced_non_class_or_alias_template_specialization_type`.

hokein committed:
Fix the link in release note.

hokein committed:
address comment.

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -96,6 +96,10 @@ C++20 Feature Support
   behavior can use the flag '-Xclang -fno-skip-odr-check-in-gmf'.
   (`#79240 <https://github.com/llvm/llvm-project/issues/79240>`_).
 
+- Initial support for class template argument deduciton (CTAD) for type alias
```
```suggestion
- Initial support for class template argument deduction (CTAD) for type alias
```

hokein replied:
Done.

erichkeane reviewed (commented):
I'm happy when Corentin is, so he can approve when he's ready.

cor3ntin commented:
Can you mark the status of the paper in cxx_status.html as partial (with a `<details></details>` tag explaining what is missing)?
Can you update the commit message / changelog to refer to the paper number?

Thanks

h-vetinari commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -96,6 +96,10 @@ C++20 Feature Support
   (`#79240 <https://github.com/llvm/llvm-project/issues/79240>`_).
 
+- Initial support for class template argument deduciton (CTAD) for type alias
+  templates (`P1814R0 <https://wg21.link/p1814r0>`_).
```
Probably should update to at least a `<td class="partial" align="center">Partial</td>` under
https://github.com/llvm/llvm-project/blob/e946b5a87b2db307da076093d0a9a72ecb4ec089/clang/www/cxx_status.html#L843-L847

hokein replied:
Done.

hokein committed:
- rebase to main
- update err_deduced_non_class_template_specialization_type to include alias templates
- update cxx_status.html
- fix a typo in ReleaseNotes

hokein committed:
clang-format

Endilll reviewed (commented):
`Sema.h` changes look good.

hokein committed:
another clang-format fix

cor3ntin reviewed (approved):
LGTM
Thanks for working on that!

hokein commented:
Thank you for all the reviews, I will wait 1 or 2 days before merging it in case anyone has more comments.

cor3ntin commented:
> Thank you for all the reviews, I will wait 1 or 2 days before merging it in case anyone has more comments.

Sure!
Do you plan to keep working on completing this feature?

hokein commented:
> > Thank you for all the reviews, I will wait 1 or 2 days before merging it in case anyone has more comments.
> 
> Sure! Do you plan to keep working on completing this feature?

Yeah, I plan to complete this feature before the next clang release. If anyone is willing to help, that would be even greater.

BTW, should we close https://github.com/llvm/llvm-project/issues/54051 with this PR? or should we keep it open until the feature is fully implemented? The `is_deducible` part is a major missing piece, I think we can close #54051 and open a separate issue to track the missing piece, WDYT?

cor3ntin commented:
> BTW, should we close #54051 with this PR? or should we keep it open until the feature is fully implemented? The `is_deducible` part is a major missing piece, I think we can close #54051 and open a separate issue to track the missing piece, WDYT?

Sounds good
